bn-apps/cordapp-updates-distribution/corda-updates-core/src/main/kotlin/net/corda/businessnetworks/cordaupdates/core/CordaMavenResolver.kt

```diff
@@ -0,0 +1,248 @@
+            downloadVersion(artifactVersion.toString(), configProps).versions.single()
+        }.asReversed()
+        return artifactMetadata.copy(versions = versions)
+    }
+
+    /**
+     * Downloads a specified version of a CorDapp if it's missing in the local repo.
+     *
+     * @param coordinates full maven coordinates, i.e. "net.corda:corda-finance:3.2".
+     * @param configProps will be passed as custom Maven Resolver session properties
+     *
+     * @return artifact metadata with a single version
+     *
+     * @throws [ResourceNotFoundException] if the the specified artifact has not been found in the remote repository
+     * @throws [ResourceTransferException] if the remote repository is unreachable
+     */
+    fun downloadVersion(coordinates : String,
+                        configProps : Map<String, Any> = mapOf()) : ArtifactMetadata {
+        val session = createRepositorySession(configProps)
+        val aetherArtifact = DefaultArtifact(coordinates)
+        val artifactRequest = ArtifactRequest(aetherArtifact, listOf(remoteRepository), null)
+        val resolutionResult : ArtifactResult
+        try {
+            resolutionResult = repositorySystem.resolveArtifact(session, artifactRequest)!!
+        } catch (ex : RepositoryException) {
+            when (ex.cause) {
+                is ArtifactNotFoundException -> throw ResourceNotFoundException(coordinates, ex)
+                is ArtifactTransferException -> throw ResourceTransferException(coordinates, ex)
+                else -> throw CordaMavenResolverException("Error while resolving the artifact $coordinates", ex)
+            }
+        }
+        return ArtifactMetadata(aetherArtifact.groupId,
+                aetherArtifact.artifactId,
+                aetherArtifact.classifier,
+                aetherArtifact.extension,
+                listOf(VersionMetadata(aetherArtifact.version, resolutionResult.repository is LocalRepository)))
+    }
+
+    private fun createRepositorySession(additionalConfigProperties : Map<String, Any>) : DefaultRepositorySystemSession {
+        val session = newSession()!!
+        (configProperties + additionalConfigProperties).forEach {
+            session.setConfigProperty(it.key, it.value)
+        }
+        if (repositoryListener != null) session.repositoryListener = repositoryListener
+        if (transferListener != null) session.transferListener = transferListener
+        session.localRepositoryManager = SimpleLocalRepositoryManagerFactory().newInstance(session, LocalRepository(localRepoPath))
+        return session
+    }
+}
+
+/**
+ * Representation of an artifact and its versions
+ */
+data class ArtifactMetadata(val group : String, val name : String, val classifier : String = "", val extension : String = "jar", val versions : List<VersionMetadata> = listOf()) {
+    fun toMavenArtifacts() = versions.map { DefaultArtifact(group, name, classifier, extension, it.version) }
```

Comment:
Why not just embed Artifact directly?

Reply:
Makes sense. Changed to embed DefaultArtifact directly.